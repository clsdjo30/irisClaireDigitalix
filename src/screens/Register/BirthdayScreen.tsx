import React from 'react';
import { StyleSheet, Text, View, SafeAreaView, Button, Platform, TextInput, TouchableOpacity, } from 'react-native';
import { Icon } from '@rneui/base'
import { StackScreenProps } from '@react-navigation/stack';
import { useUserStore } from '../../utils/hooks/useUserStore';
import DateTimePicker from '@react-native-community/datetimepicker';
import { colors } from '../../theme'


const BirthdayScreen: React.FC<StackScreenProps<any>> = ({ navigation }) => {

    const [user, setUser] = useUserStore();
    const [date, setDate] = React.useState(new Date());
    const [mode, setMode] = React.useState('date');
    const [show, setShow] = React.useState(false);

    const onChange = (event: any, selectedDate: any) => {
        const currentDate = selectedDate || date;
        setShow(Platform.OS === 'ios');
        setDate(currentDate);
        setUser({ ...user, birthday: currentDate.toLocaleDateString() })
    };

    const showMode = (currentMode: React.SetStateAction<string>) => {
        
        setShow(true);
        setMode(currentMode);
    };

    const showDatepicker = () => {
        showMode('date');
    };

    return (

        <SafeAreaView style={styles.container}>

            <Text style={styles.contentTitle}>Quelle est votre date de naissance ?</Text>
            <View style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'center' }}>

                <View >
                    <TouchableOpacity onPress={showDatepicker} style={styles.control}>
                        <Icon 
                        name="birthday-cake" 
                        type='font-awesome'
                        size={28} 
                        color= {colors.palette.lightgold}
                        style={styles.icon} 
                        />
                        <TextInput
                            style={styles.input}
                            onFocus={showDatepicker}
                            placeholder="Date de naissance"
                            placeholderTextColor={'#8ca0d7'}
                            value={user.birthday}
                            editable={false}
                        />
                        {show && (
                            <DateTimePicker
                                testID="dateTimePicker"
                                value={date}
                                mode={mode}
                                is24Hour={true}
                                display="default"
                                onChange={onChange}
                            />)}

                    </TouchableOpacity>
                    <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('Sign Up')}>
                        <Text style={styles.buttonText}>Suivant</Text>
                    </TouchableOpacity>
                </View>
            </View>
        </SafeAreaView>
    );
};





const styles = StyleSheet.create({
    container: {
        flex: 1,
        height: '100%',
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: colors.palette.purple600,
    },
    button: {
        position: 'relative',
        top: 50,
        width: 300,
        backgroundColor: "#CBA135",
        marginTop: 10,
        borderRadius: 16,
    },
    buttonText: {
        textAlign: "center",
        padding: 3,
        fontFamily: "oswaldMedium",
        fontSize: 14,
        color: colors.palette.ivory,
    },
    input: {
        color: '#8ca0d7',
        fontSize: 14,
        paddingLeft: 20,
    },
    control: {
        width: 300,
        flexDirection: 'row',
        alignItems: 'center',
        paddingBottom: 10,
        backgroundColor: colors.palette.ivory,
        borderRadius: 10,
        padding: 10,
        borderBottomWidth: 1,
        borderLeftWidth: 1,
        borderLeftColor: colors.palette.darkgold,
        borderBottomColor: colors.palette.darkgold,
    },
    icon: {
        marginRight: 20,
        color: '#FFD700',
    },
    contentTitle: {
        fontFamily: "mulishRegular",
        fontSize: 18,
        color: colors.text,
        marginBottom: 20
    },
});

export default BirthdayScreen;