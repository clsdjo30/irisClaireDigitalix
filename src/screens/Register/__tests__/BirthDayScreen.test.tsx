import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { goToSignUp } from '../BirthdayScreen';
import BirthdayScreen from '../BirthdayScreen';

// Mock de useUserStore
jest.mock('../../../hooks/useUserStore');
const mockUseUserStore = require('../../../hooks/useUserStore').useUserStore;

// Mock de la navigation
const mockNavigation: any = {
    navigate: jest.fn(),
};
const mockRoute: any = {
    key: 'MockKey',
    name: 'MockName',
};
const mockProps = {
    navigation: mockNavigation,
    route: mockRoute,
};

// Mock the Icon component
jest.mock('@rneui/base', () => ({
    Icon: () => 'Icon',
}));

describe('BirthDayScreen component', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        const date = new Date(2000, 0, 1); // 1er janvier 2000
        mockUseUserStore.mockReturnValue([
            { birthday: date.toISOString() },
            jest.fn(),
        ]);
    });

    it('renders correctly and matches snapshot', () => {
        const tree = render(<BirthdayScreen {...mockProps} />);
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly', () => {
        const { getByTestId } = render(<BirthdayScreen {...mockProps} />);
        expect(getByTestId('birthday-screen')).toBeTruthy();
    });

    it('contains the key elements', () => {
        const { getByText, getByPlaceholderText } = render(<BirthdayScreen {...mockProps} />);
        expect(getByText('Quelle est votre date de naissance ?')).toBeTruthy();
        expect(getByPlaceholderText('Date de naissance')).toBeTruthy();
    });

    it('should not call navigation.navigate when "S\'inscrire" button is pressed and the user is under 18', () => {
        mockUseUserStore.mockReturnValueOnce([
            { birthday: new Date().toLocaleDateString() }, // setting the birthday to today's date
            jest.fn(),
        ]);

        const { getByText } = render(<BirthdayScreen {...mockProps} />);
        fireEvent.press(getByText("S'inscrire"));

        expect(mockNavigation.navigate).not.toHaveBeenCalled();
    });

    it('should display an error message if the user is under 18', () => {
        mockUseUserStore.mockReturnValueOnce([
            { birthday: new Date().toLocaleDateString() }, // setting the birthday to today's date
            jest.fn(),
        ]);

        const { getByText } = render(<BirthdayScreen {...mockProps} />);
        fireEvent.press(getByText("S'inscrire"));

        expect(getByText("Vous devez avoir plus de 18 ans pour vous inscrire")).toBeTruthy();
    });

    it('should set error if birthday is not defined', () => {
        mockUseUserStore.mockReturnValueOnce([
            { birthday: '' }, // Mocking an empty birthday
            jest.fn(),
        ]);

        const { getByText } = render(<BirthdayScreen {...mockProps} />);
        fireEvent.press(getByText("S'inscrire"));

        expect(getByText("Votre date de naissance ne peut pas Ãªtre vide")).toBeTruthy();
        expect(mockNavigation.navigate).not.toHaveBeenCalled();
    });

    it('should set error if user is under 18', () => {
        const currentDate = new Date();
        const seventeenYearsAgo = new Date(currentDate.getFullYear() - 17, currentDate.getMonth(), currentDate.getDate()).toLocaleDateString();

        mockUseUserStore.mockReturnValueOnce([
            { birthday: seventeenYearsAgo }, // Mocking a birthday that makes user 17 years old
            jest.fn(),
        ]);

        const { getByText } = render(<BirthdayScreen {...mockProps} />);
        fireEvent.press(getByText("S'inscrire"));

        expect(getByText("Vous devez avoir plus de 18 ans pour vous inscrire")).toBeTruthy();
        expect(mockNavigation.navigate).not.toHaveBeenCalled();
    });

    // test goToSignUp function

    

   
});

