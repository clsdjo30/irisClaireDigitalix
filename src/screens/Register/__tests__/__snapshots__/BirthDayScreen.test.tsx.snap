// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BirthDayScreen component renders correctly and matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2F2A5A",
      "flex": 1,
      "height": "100%",
      "justifyContent": "center",
    }
  }
  testID="birthday-screen"
>
  <Text
    style={
      {
        "color": "#eeedf6",
        "fontFamily": "mulishRegular",
        "fontSize": 18,
        "marginBottom": 20,
      }
    }
  >
    Quelle est votre date de naissance ?
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#eeedf6",
            "borderBottomColor": "#bc9531",
            "borderBottomWidth": 1,
            "borderLeftColor": "#bc9531",
            "borderLeftWidth": 1,
            "borderRadius": 10,
            "flexDirection": "row",
            "padding": 10,
            "paddingBottom": 10,
            "width": 300,
          }
        }
      >
        Icon
        <TextInput
          editable={false}
          onFocus={[Function]}
          placeholder="Date de naissance"
          placeholderTextColor="#B7B7E1"
          style={
            {
              "color": "#8ca0d7",
              "fontSize": 14,
              "paddingLeft": 20,
            }
          }
          value="1999-12-31T23:00:00.000Z"
        />
      </View>
      <View
        style={
          {
            "marginTop": 60,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 16,
                "marginBottom": 10,
                "paddingVertical": 5,
              },
              {
                "backgroundColor": "#FF7648",
                "width": 576.9230769230769,
              },
            ]
          }
          testID="goToSignUp"
        >
          <Text
            style={
              [
                {
                  "fontFamily": "mulishBold",
                  "fontSize": 14,
                },
                {
                  "color": "#eeedf6",
                },
              ]
            }
          >
            S'inscrire
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
